fleetx/models/gpt_model/utils.py

```diff
@@ -0,0 +1,369 @@
+from paddle import framework
+from paddle.nn import functional as F
+from paddle.autograd import PyLayer
+from paddle.fluid import core
+from paddle.fluid.dygraph.layers import Layer
+from paddle.distributed import fleet
+from paddle.distributed.fleet.base import topology as tp
+from paddle.distributed.fleet.meta_parallel import get_rng_state_tracker
+
+import numpy as np
+
+
+
+####################################################
+#                                                  #
+#        Distributed Communication Operator        #
+#                                                  #
+####################################################
+
+class ScatterOp(PyLayer):
+    # input shape: [s, b, h], n is mp parallelism
+    # after forward shape: [s/n, b, h]
+    @staticmethod
+    def forward(self, input):
+        hcg = fleet.get_hybrid_communicate_group()
+        group = hcg.get_model_parallel_group()
+        parallelism = group.nranks
+        rank = group.rank
+        assert input.shape[0] % parallelism == 0, \
+            "Input sequence length {0} can't be divided exactly \
+             by sequence parallelism {1}".format(input.shape[0], parallelism)
+        input = paddle.split(input, num_or_sections=parallelism, axis=0)[rank]
+        return input
+
+    @staticmethod
+    def backward(self, grad):
+        hcg = fleet.get_hybrid_communicate_group()
+        group = hcg.get_model_parallel_group()
+        parallelism = group.nranks
+        output_shape = grad.shape
+        output_shape[0] = output_shape[0] * parallelism
+        output = paddle.empty(shape=output_shape, dtype=grad.dtype)
+        group.process_group.all_gather(grad, output).wait()
+        return output
+
+class GatherOp(PyLayer):
+    # input shape: [s/n, b, h], n is mp parallelism
+    # after forward shape: [s, b, h]
+    @staticmethod
+    def forward(self, input):
+        hcg = fleet.get_hybrid_communicate_group()
+        group = hcg.get_model_parallel_group()
+        parallelism = group.nranks
+        output_shape = input.shape
+        output_shape[0] = output_shape[0] * parallelism
+        output = paddle.empty(shape=output_shape, dtype=input.dtype)
+        group.process_group.all_gather(input, output).wait()
+        return output
+
+    @staticmethod
+    def backward(self, grad):
+        hcg = fleet.get_hybrid_communicate_group()
+        group = hcg.get_model_parallel_group()
+        parallelism = group.nranks
+        rank = group.rank
+        assert grad.shape[0] % parallelism == 0, \
+            "Input sequence length {0} can't be divided exactly \
+             by sequence parallelism {1}".format(grad.shape[0], parallelism)
+        input = paddle.split(grad, num_or_sections=parallelism, axis=0)[rank]
+        return input
+
+# All gather along the first dim during forward pass
+# All reduce and scatter along the first dim during backward pass
+class AllGatherOp(PyLayer):
+    # input shape: [s/n, b, h], n is mp parallelism
```

Comment:
Duplicate code with `ReduceScatterOp`. Can use two functions to simplify the codes.

```Python
def all_gather(...): pass

def reduce_scatter(...): pass

class ReduceScatterOp(PyLayer):
    @staticmethod
    def forward(ctx, ...):
        return reduce_scatter(...)

    @staticmethod
    def backward(ctx, ...):
         return all_gather(...)

class AllGatherOp(PyLayer): ...
```